test/src/net.rs

```diff
@@ -1,111 +1,91 @@
-use crate::specs::TestProtocol;
-use crate::utils::{temp_path, wait_until};
-use crate::{Node, Setup};
+use crate::global::VENDOR_PATH;
+use crate::utils::{find_available_port, temp_path, wait_until};
+use crate::Node;
 use ckb_app_config::NetworkConfig;
+use ckb_chain_spec::consensus::Consensus;
+
 use ckb_channel::{self as channel, Receiver, RecvTimeoutError, Sender};
 use ckb_network::{
     bytes::Bytes, CKBProtocol, CKBProtocolContext, CKBProtocolHandler, DefaultExitHandler,
-    NetworkController, NetworkService, NetworkState, PeerIndex, ProtocolId,
+    NetworkController, NetworkService, NetworkState, PeerIndex, ProtocolId, SupportProtocols,
 };
-use ckb_types::core::{BlockNumber, BlockView};
-use std::collections::HashSet;
+
 use std::path::PathBuf;
-use std::sync::{
-    atomic::{AtomicU16, Ordering},
-    Arc,
-};
+use std::sync::Arc;
 use std::time::Duration;
 
 pub type NetMessage = (PeerIndex, ProtocolId, Bytes);
 
 pub struct Net {
-    pub nodes: Vec<Node>,
-    controller: Option<(NetworkController, Receiver<NetMessage>)>,
+    consensus: Consensus,
+    protocols: Vec<SupportProtocols>,
     p2p_port: u16,
-    setup: Setup,
     working_dir: String,
-    vendor_dir: PathBuf,
+    controller: Option<(NetworkController, Receiver<NetMessage>)>,
 }
 
 impl Net {
-    pub fn new(
-        binary: &str,
-        start_port: Arc<AtomicU16>,
-        vendor_dir: PathBuf,
-        setup: Setup,
-        case_name: &str,
-    ) -> Self {
-        let p2p_port = start_port.fetch_add(1, Ordering::SeqCst);
-        let nodes: Vec<Node> = (0..setup.num_nodes)
-            .enumerate()
-            .map(|(index, _)| {
-                let node_index = "node".to_owned() + &index.to_string();
-                let p2p_port = start_port.fetch_add(1, Ordering::SeqCst);
-                let rpc_port = start_port.fetch_add(1, Ordering::SeqCst);
-                Node::new(binary, p2p_port, rpc_port, case_name, &node_index)
-            })
-            .collect();
-
+    pub fn new(spec_name: &str, consensus: Consensus, protocols: Vec<SupportProtocols>) -> Self {
+        assert!(
+            !protocols.is_empty(),
+            "Net cannot initialize with empty protocols"
+        );
+        let p2p_port = find_available_port();
+        let working_dir = temp_path(spec_name, "net");
         Self {
-            nodes,
-            controller: None,
+            consensus,
+            protocols,
             p2p_port,
-            setup,
-            working_dir: temp_path(case_name, "net"),
-            vendor_dir,
+            working_dir,
+            controller: None,
         }
     }
 
     pub fn working_dir(&self) -> &str {
         &self.working_dir
     }
 
-    pub fn vendor_dir(&self) -> &PathBuf {
-        &self.vendor_dir
+    pub fn vendor_dir(&self) -> PathBuf {
+        let vendor_path = VENDOR_PATH.lock();
+        (*vendor_path).clone()
```

Comment:
how about use `global::vendor()`?

Reply:
The function `fn Net::vendor()` is useless, so I have removed it. Thanks.